### PR TITLE
Synchronize progress table scrolling

### PR DESCRIPTION
**Pull request by cforkish**
while investigating [LP-1933] i discovered a number of bugs around vertical scrolling when switching between lesson and level table views with detail rows expanded. the solution required overriding more of reactabular's internal functionality and therefore is definitely on the hacky side. let this serve as another +1 for upgrading react and getting off reactabular ASAP.

BEFORE:
Screen Shot 2021-06-01 at 5 08 56 PM

one feature that fell out of this work is that we are now able to preserve the vertical scroll position when switching view types (or more accurately, we restore it after the new instance renders, which is not ideal, but seemed like a better solution than simply resetting it to 0 for both student list and content view).

details of the change are in the comments.

[LP-1933]: https://codedotorg.atlassian.net/browse/LP-1933

### CODE REVIEW

File: apps/src/templates/sectionProgress/progressTables/ProgressTableContainer.story.jsx [hunk 1]
```diff
@@ -12,86 +14,116 @@ import {createStore, wrapTable} from '../sectionProgressTestHelpers';
...
+      <div
+        className="main"
+        style={{
+          marginLeft: 80,
+          width: 970,
+          display: 'block',
+          backgroundColor: '#ffffff'
+        }}
+      >
+        <SectionProgressToggle />
```
Comment (cforkish) [inline apps/src/templates/sectionProgress/progressTables/ProgressTableContainer.story.jsx:32]: testing this work required being able to toggle between view types, hence this update to the storybook. this update is not directly related to the purpose of this PR.

File: apps/src/templates/sectionProgress/progressTables/ProgressTableView.jsx [hunk 2]
```diff
@@ -136,29 +137,72 @@ class ProgressTableView extends React.Component {
...
+    this.contentView?.bodyComponent.setState({
+      amountOfRowsToRender: initialRows
+    });
+    this.syncScrollTop();
   }
 
   onScroll(e) {
     this.studentList.body.scrollTop = e.target.scrollTop;
     this.contentView.header.scrollLeft = e.target.scrollLeft;
+    this.scrollTop = e.target.scrollTop;
```
Comment (cforkish) [inline apps/src/templates/sectionProgress/progressTables/ProgressTableView.jsx:159]: the crux of this change is that we now store the target `scrollTop` during scrolling so we can force synchronization when scrolling finishes, view type changes, or rows are expanded/collapsed.

File: apps/src/templates/sectionProgress/progressTables/ProgressTableView.jsx [hunk 3]
```diff
@@ -136,29 +137,72 @@ class ProgressTableView extends React.Component {
...
+   * Simply setting the `scrollTop` value on each of our table components
+   * sometimes leads to the internal state of the components (and resulting
+   * vertical scroll positions) getting out of sync.
+   *
+   * To work around this, we copy the implementation of reactabular's internal
+   * `VirtualizedBody.getRef().scrollTo()` to make sure the state gets updated
+   * appropriately. However, `scrollTo` accepts a row index, whereas we need to
+   * set the `scrollTop` value directly, hence this custom implementation.
+   */
+  setScrollState(table) {
```
Comment (cforkish) [inline apps/src/templates/sectionProgress/progressTables/ProgressTableView.jsx:195]: see reference implementation [here](https://github.com/reactabular/reactabular/blob/ee7d72fb9b23d7358e0ab438b1041ae7e58fa9b1/packages/reactabular-virtualized/src/body.js#L156)

File: apps/src/templates/sectionProgress/progressTables/ProgressTableView.jsx [hunk 4]
```diff
@@ -271,12 +316,17 @@ class ProgressTableView extends React.Component {
...
+    // table components when the view changes
+    const key = this.props.currentView;
+
     return (
       // outer div contains both table and legend
       <div>
         <div style={styles.container} className="progress-table">
           <div style={styles.studentList} className="student-list">
             <ProgressTableStudentList
+              key={key}
```
Comment (cforkish) [inline apps/src/templates/sectionProgress/progressTables/ProgressTableView.jsx:329]: re-instantiating the content view without re-instantiating the student list was at the core of the problems i was seeing -- not only did that mean resetting the vertical scroll of the content view and not the student list, but also resulted in their internal states being out of sync.

**Review by maureensturgeon — APPROVED** [review]
Thanks for digging into this!
